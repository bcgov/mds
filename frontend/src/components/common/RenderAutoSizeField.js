import React from "react";
import PropTypes from "prop-types";
import { Form, Input } from "antd";

/**
 * @constant  RenderAutoSizeField - Ant Design `Input` autosize component for redux-form. (useful for notes/description)
 */

const propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  input: PropTypes.objectOf(PropTypes.any).isRequired,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  meta: PropTypes.objectOf(PropTypes.any).isRequired,
};

const defaultProps = {
  placeholder: "",
  label: "",
};

const { TextArea } = Input;
const RenderAutoSizeField = (props) => (
  <Form.Item
    label={props.label}
    placeholder={props.placeholder}
    validateStatus={
      props.meta.touched ? (props.meta.error && "error") || (props.meta.warning && "warning") : ""
    }
    help={
      props.meta.touched &&
      ((props.meta.error && <span>{props.meta.error}</span>) ||
        (props.meta.warning && <span>{props.meta.warning}</span>))
    }
  >
    <TextArea id={props.id} {...props.input} autosize placeholder={props.placeholder} />
  </Form.Item>
);

RenderAutoSizeField.propTypes = propTypes;
RenderAutoSizeField.defaultProps = defaultProps;

export default RenderAutoSizeField;
