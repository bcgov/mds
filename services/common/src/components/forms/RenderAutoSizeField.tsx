import React, { FC } from "react";
import { Input, Form, Row } from "antd";
import { BaseInputProps, BaseViewInput, getFormItemLabel } from "./BaseInput";
import { FormConsumer, IFormContext } from "./FormWrapper";

/**
 * @constant  RenderAutoSizeField - Ant Design `Input` autosize component for redux-form. (useful for notes/description)
 */

interface AutoSizeProps extends BaseInputProps {
  minRows?: number;
  maximumCharacters: number;
}

const RenderAutoSizeField: FC<AutoSizeProps> = ({
  label = "",
  labelSubtitle,
  help,
  disabled = false,
  maximumCharacters,
  minRows = 3,
  required = false,
  ...props
}) => {
  return (
    <FormConsumer>
      {(value: IFormContext) => {
        if (!value.isEditMode) {
          return <BaseViewInput value={props.input.value} label={label} />;
        }
        return (
          <Form.Item
            name={props.input.name}
            required={required}
            label={getFormItemLabel(label, required, labelSubtitle)}
            validateStatus={
              props.meta.touched
                ? (props.meta.error && "error") || (props.meta.warning && "warning")
                : ""
            }
            help={
              props.meta.touched &&
              ((props.meta.error && <span>{props.meta.error}</span>) ||
                (props.meta.warning && <span>{props.meta.warning}</span>))
            }
          >
            <>
              <Input.TextArea
                disabled={disabled}
                id={props.id}
                {...props.input}
                autoSize={{ minRows: minRows }}
                placeholder={props.placeholder}
              />
              <Row
                justify="space-between"
                className={`form-item-help ${props.input.name}-form-help`}
              >
                {help ? (
                  <span>{help}</span>
                ) : (
                  <span>{`Maximum ${maximumCharacters} characters`}</span>
                )}
                <span className="flex-end">{`${maximumCharacters -
                  props.input.value.length} / ${maximumCharacters}`}</span>
              </Row>
            </>
          </Form.Item>
        );
      }}
    </FormConsumer>
  );
};
export default RenderAutoSizeField;
