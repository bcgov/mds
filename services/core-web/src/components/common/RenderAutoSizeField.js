import React, { useState } from "react";
import PropTypes from "prop-types";
import { Form } from "@ant-design/compatible";
import "@ant-design/compatible/assets/index.css";
import { Input } from "antd";

/**
 * @constant  RenderAutoSizeField - Ant Design `Input` autosize component for redux-form. (useful for notes/description)
 */

const propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  input: PropTypes.objectOf(PropTypes.any).isRequired,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  meta: PropTypes.objectOf(PropTypes.any).isRequired,
  disabled: PropTypes.bool,
  minRows: PropTypes.number,
  maximumCharacters: PropTypes.number,
};

const defaultProps = {
  placeholder: "",
  label: "",
  disabled: false,
  minRows: 3,
  maximumCharacters: 300,
};

const RenderAutoSizeField = (props) => {
  const [remainingChars, setRemainingChars] = useState(props.maximumCharacters);
  const [value, setValue] = useState("");

  const handleTextAreaChange = (event) => {
    setValue(event.target.value);

    const input = event.target.value;
    const remaining = props.maximumCharacters - input.length;
    setRemainingChars(remaining);
  };

  return (
    <Form.Item
      label={props.label}
      placeholder={props.placeholder}
      validateStatus={
        props.meta.touched ? (props.meta.error && "error") || (props.meta.warning && "warning") : ""
      }
      help={
        props.meta.touched &&
        ((props.meta.error && <span>{props.meta.error}</span>) ||
          (props.meta.warning && <span>{props.meta.warning}</span>))
      }
    >
      <Input.TextArea
        disabled={props.disabled}
        id={props.id}
        {...props.input}
        autoSize={{ minRows: props.minRows }}
        placeholder={props.placeholder}
        onChange={handleTextAreaChange}
        value={value}
      />
      {props.maximumCharacters > 0 && (
        <div className="flex between">
          <span>{`Maximum ${props.maximumCharacters} characters`}</span>
          <span className="flex-end">{`${remainingChars} / ${props.maximumCharacters}`}</span>
        </div>
      )}
    </Form.Item>
  );
};

RenderAutoSizeField.propTypes = propTypes;
RenderAutoSizeField.defaultProps = defaultProps;

export default RenderAutoSizeField;
