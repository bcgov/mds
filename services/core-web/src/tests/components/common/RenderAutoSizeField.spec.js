import React from 'react';
import { shallow } from 'enzyme';
import RenderAutoSizeField from '@/components/common/RenderAutoSizeField';

let props = {};

const setupProps = () => {
  props = {
    id: 1,
    input: '',
    label: '',
    type: '',
    meta: {
      touched: false,
      error: false,
      warning: false,
    },
  };
}

beforeEach(() => {
  setupProps();
});

describe('RenderAutoSizeField', () => {
  it('renders properly', () => {
    const wrapper = shallow(<RenderAutoSizeField {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});