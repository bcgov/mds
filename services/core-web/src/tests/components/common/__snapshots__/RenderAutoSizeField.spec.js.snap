// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderAutoSizeField renders properly 1`] = `
<FormItem
  hasFeedback={false}
  help={false}
  label=""
  placeholder=""
  validateStatus=""
>
  <TextArea
    autoSize={
      Object {
        "minRows": 4,
      }
    }
    disabled={false}
    id={1}
    placeholder=""
  />
</FormItem>
`;
